File: code/go/ch07/src/luago/vm/inst_load.go
package vm

import . "luago/api"

// R(A), R(A+1), ..., R(A+B) := nil
func loadNil(i Instruction, vm LuaVM) {
	a, b, _ := i.ABC()
	a += 1

	vm.PushNil()
	for i := a; i <= a+b; i++ {
		vm.Copy(-1, i)
	}
	vm.Pop(1)
}

// R(A) := (bool)B; if (C) pc++
func loadBool(i Instruction, vm LuaVM) {
	a, b, c := i.ABC()
	a += 1

	vm.PushBoolean(b != 0)
	vm.Replace(a)

	if c != 0 {
		vm.AddPC(1)
	}
}

// R(A) := Kst(Bx)
func loadK(i Instruction, vm LuaVM) {
	a, bx := i.ABx()
	a += 1

	vm.GetConst(bx)
	vm.Replace(a)
}

// R(A) := Kst(extra arg)
func loadKx(i Instruction, vm LuaVM) {
	a, _ := i.ABx()
	a += 1
	ax := Instruction(vm.Fetch()).Ax()

	//vm.CheckStack(1)
	vm.GetConst(ax)
	vm.Replace(a)
}

File: code/csharp/ch07/luavm/vm/InstLoad.cs
using LuaVm = luavm.api.LuaState;

namespace luavm.vm
{
    public class InstLoad
    {
        internal static void loadNil(Instruction i, LuaVm vm)
        {
            var ab_ = i.AsBx();
            var a = ab_.Item1 + 1;
            var b = ab_.Item2;

            vm.PushNil();
            for (var l = a; l <= a + b; l++)
            {
                vm.Copy(-1, l);
            }

            vm.Pop(1);
        }

        internal static void loadBool(Instruction i, LuaVm vm)
        {
            var abc = i.ABC();
            var a = abc.Item1 + 1;
            var b = abc.Item2;
            var c = abc.Item3;

            vm.PushBoolean(b != 0);
            vm.Replace(a);
            if (c != 0)
            {
                vm.AddPC(1);
            }
        }

        internal static void loadK(Instruction i, LuaVm vm)
        {
            var aBx = i.ABx();
            var a = aBx.Item1 + 1;
            var bx = aBx.Item2;

            vm.GetConst(bx);
            vm.Replace(a);
        }

        internal static void loadKx(Instruction i, LuaVm vm)
        {
            var aBx = i.ABx();
            var a = aBx.Item1 + 1;
            var ax = new Instruction(vm.Fetch()).Ax();

            vm.GetConst(ax);
            vm.Replace(a);
        }
    }
}